plugins/modules/xen_orchestra_instance.py

```diff
@@ -0,0 +1,326 @@
+
+        if 'error' in answer:
+            # VM is not paused, suspended or running
+            if answer['error']['code'] == VM_STATE_ERROR:
+                return False
+            raise self.module.fail_json(
+                'Could not stop vm: {0}'.format(answer['error']))
+
+        return answer['result']
+
+    def start_vm(self, vm_uid):
+        answer = self.call('vm.start', {'id': vm_uid})
+
+        if 'error' in answer:
+            # VM is already started, nothing to do
+            if answer['error']['code'] == VM_STATE_ERROR:
+                return False
+            raise self.module.fail_json(
+                'Could not start vm: {0}'.format(answer['error']))
+
+        return answer['result']
+
+    def delete_vm(self, vm_uid):
+        answer = self.call('vm.delete', {'id': vm_uid})
+
+        if 'error' in answer:
+            if answer['error']['code'] == OBJECT_NOT_FOUND:
+                return False
+            raise self.module.fail_json(
+                'Could not delete vm: {0}'.format(answer['error']))
+
+        return answer['result']
+
+
+def main():
+    module_args = dict(
+        api_host=dict(type='str', required=True),
+        user=dict(type='str', required=True),
+        password=dict(type='str', required=True, no_log=True),
+        validate_certs=dict(type='bool', default=True),
+        use_tls=dict(type='bool', default=True),
+        template=dict(type='str'),
+        vm_uid=dict(type='str'),
+        label=dict(type='str'),
+        description=dict(type='str'),
+        boot_after_create=dict(type='bool', default=False),
+        state=dict(default='present', choices=['present', 'absent', 'stopped', 'started', 'restarted']),
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        required_if=[
+            ('state', 'present', ['template', 'label']),
+            ('state', 'absent', ('vm_uid',)),
+            ('state', 'started', ('vm_uid',)),
+            ('state', 'restarted', ('vm_uid',)),
+            ('state', 'stopped', ('vm_uid',)),
+        ],
+    )
+
+    if HAS_WEBSOCKET is False:
+        module.fail_json(msg=missing_required_lib('websocket-client'), exception=WEBSOCKET_IMP_ERR)
+
+    xen_orchestra = XenOrchestra(module)
+
+    state = module.params['state']
+    vm_uid = module.params['vm_uid']
+
+    if state == 'stopped':
+        result = xen_orchestra.stop_vm(vm_uid)
+        module.exit_json(changed=result)
+
+    if state == 'started':
+        result = xen_orchestra.start_vm(vm_uid)
+        module.exit_json(changed=result)
+
+    if state == 'restarted':
+        result = xen_orchestra.restart_vm(vm_uid)
+        module.exit_json(changed=result)
+
+    if state == 'absent':
+        result = xen_orchestra.delete_vm(vm_uid)
+        module.exit_json(changed=result)
+
+    if state == 'present':
+        result = xen_orchestra.create_vm()
+        module.exit_json(changed=False, vm_uid=result)
```

Comment:
Is `changed=False` intentional here? I guess it should be `changed=result` as above.

Reply:
Not sure what's the best option is here. The two expected params are a name and the id of a template

If the template id is not found then an error is raised. This path is handled.
Otherwise we will create a new vm each time. I guess we should always set `changed` to `True`?

Reply:
If there is a change, `changed` should always be set to `True`.

Reply:
This is still open.